integration-tests/checkstyle-10.12.4.sh

```diff
@@ -0,0 +1,162 @@
+
+test_name="$(basename "${0}" .sh)"
+project=checkstyle
+repository=git@github.com:checkstyle/checkstyle.git
+revision=checkstyle-10.12.4
+
+if [ "${#}" -gt 1 ] || [[ ${1:---sync} != '--sync' ]]; then
+  echo "Usage: ${0} [--sync]"
+  exit 1
+fi
+do_sync="${1:-}"
+
+# XXX: Configure Renovate to manage the AssertJ and fmt-maven-plugin versions
+# declared here.
+shared_build_flags="
+  -Perror-prone-compile,error-prone-test-compile
+  -Dassertj.version=3.24.2
+  -Derror-prone.version=$(
+    mvn -f "${error_prone_support_root}" help:evaluate -Dexpression=version.error-prone -q -DforceStdout
+  )
+  -Derror-prone-support.version=$(
+    mvn -f "${error_prone_support_root}" help:evaluate -Dexpression=project.version -q -DforceStdout
+  )
+  -Dfmt.version=2.21.1
+  -DadditionalSourceDirectories=\${project.basedir}\${file.separator}src\${file.separator}it\${file.separator}java,\${project.basedir}\${file.separator}src\${file.separator}xdocs-examples\${file.separator}java
+  "
+
+error_prone_shared_flags='-XepExcludedPaths:(\Q${project.basedir}${file.separator}src${file.separator}\E(it|test|xdocs-examples)\Q${file.separator}resources\E|\Q${project.build.directory}${file.separator}\E).*'
+
+error_prone_patch_flags="${error_prone_shared_flags} -XepPatchLocation:IN_PLACE -XepPatchChecks:$(
+  find "${error_prone_support_root}" -path "*/META-INF/services/com.google.errorprone.bugpatterns.BugChecker" \
+    | xargs grep -hoP '[^.]+$' \
+    | paste -s -d ','
+)"
+
+error_prone_validation_flags="${error_prone_shared_flags} -XepDisableAllChecks $(
+  find "${error_prone_support_root}" -path "*/META-INF/services/com.google.errorprone.bugpatterns.BugChecker" \
+    | xargs grep -hoP '[^.]+$' \
+    | sed -r 's,(.*),-Xep:\1:WARN,' \
+    | paste -s -d ' '
+)"
+
+validation_log_file="$(mktemp)"
+trap 'rm -rf -- "${validation_log_file}"' INT TERM HUP EXIT
+
+echo "Shared build flags: ${shared_build_flags}"
+echo "Error Prone patch flags: ${error_prone_patch_flags}"
+echo "Error Prone validation flags: ${error_prone_validation_flags}"
+
+mkdir -p "${repos_root}"
+
+# Make sure that the targeted tag of the project's Git repository is checked
+# out.
+project_root="${repos_root}/${project}"
+if [ ! -d "${project_root}" ]; then
+  # The repository has not yet been cloned; create a shallow clone.
+  git clone --branch "${revision}" --depth 1 "${repository}" "${project_root}"
+else
+  # The repository does already appear to exist. Try to check out the requested
+  # tag if possible, and fetch it otherwise.
+  #
+  # Under certain circumstances this does not cause the relevant tag to be
+  # created, so if necessary we manually create it.
+  git -C "${project_root}" checkout --force "${revision}" 2>/dev/null \
+    || (
+         git -C "${project_root}" fetch --depth 1 "${repository}" "${revision}" \
+           && git -C "${project_root}" checkout --force FETCH_HEAD \
+           && git -C "${project_root}" tag "${revision}" || true
+       )
+fi
```

Comment:
On GitHub Actions there'll always be a new checkout, but for local development it's very convenient if repeated script runs don't trigger new fetch/clone operations.